### DISCUSSION TITLE
Token for EOF

### DISCUSSION THREAD

**Discussion by NeoCortex97** [lark-parser/lark#1041] [Q&A] [answered]
How can i specify the end of the code in my grammar?
I am building a language that needs to be processed line by line.

I would like to make a line either statements with a newline at the end, or statements and the end of file.

**Comment by erezsh** [comment, marked as answer]
There is a branch of Lark the implements the end symbol as `$`: https://github.com/lark-parser/lark/tree/end_symbol_2021

We haven't merged it in, because for some grammars, which are hard to detect automatically, the parser might go into an infinite loop. But it's pretty stable per grammar, so if you get it working for yours, it's probably fine.

However,

In practice I think the easiest choice is to only check for newlines, and always add a newline at the end of your input, before sending it to the parser.

**Comment by NeoCortex97** [comment]
I tried specifying lalr as a parser and it started doing what i wanted
